lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -1722,6 +1722,61 @@ void swift::visitExtendedGuaranteedForwardingPhiBaseValuePairs(
   }
 }
 
+/// If \p instruction forwards guaranteed values to its results, visit each
+/// forwarded operand. The visitor must check whether the forwarded value is
+/// guaranteed.
+///
+/// Return true \p visitOperand was called at least once.
+///
+/// \p visitOperand should always recheck for Guaranteed owernship if it
+/// matters, in case a cast forwards a trivial type to a nontrivial type.
+///
+/// This intentionally does not handle phis, which require recursive traversal
+/// to determine `isGuaranteedForwardingPhi`.
+bool swift::visitForwardedGuaranteedOperands(
+  SILValue value, function_ref<void(Operand *)> visitOperand) {
+
+  assert(!SILArgument::asPhi(value) && "phis are handled separately");
+
+  if (auto *termResult = SILArgument::isTerminatorResult(value)) {
+    if (auto *oper = termResult->forwardedTerminatorResultOperand()) {
+      visitOperand(oper);
+      return true;
+    }
+    return false;
+  }
+  auto *inst = value->getDefiningInstruction();
+  if (!inst)
+    return false;
+
+  // Bypass conversions that produce a guarantee value out of thin air.
+  if (inst->getNumRealOperands() == 0) {
+    return false;
+  }
+  if (isa<FirstArgOwnershipForwardingSingleValueInst>(inst)
+      || isa<OwnershipForwardingConversionInst>(inst)
+      || isa<OwnershipForwardingSelectEnumInstBase>(inst)
+      || isa<OwnershipForwardingMultipleValueInstruction>(inst)
+      || isa<MoveOnlyWrapperToCopyableValueInst>(inst)
+      || isa<CopyableToMoveOnlyWrapperValueInst>(inst)) {
+    assert(inst->getNumRealOperands() == 1
+           && "forwarding instructions must have a single real operand");
+    assert(!isa<SingleValueInstruction>(inst)
+           || !BorrowedValue(cast<SingleValueInstruction>(inst))
+                  && "forwarded operand cannot begin a borrow scope");
+    visitOperand(&inst->getOperandRef(0));
+    return true;
+  }
+  if (isa<AllArgOwnershipForwardingSingleValueInst>(inst)) {
+    assert(inst->getNumOperands() > 0 && "checked above");
```

Comment:
I added
`assert(inst->getNumOperands() == getNumRealOperands() && "mixin invariant");`
since the following code makes that assumption and I don't have time to wade into the deeper problem now. The rest of the code base just makes these assumptions without even bothering to check. It's quite possible that most uses of the `getAllOperands` family of APIs are incorrect--and there are innumerable uses of these APIs.

In this limited case, we make the assumption that structs and tuples can't refer to an opened type. I can't come up with a counter-example easily. Maybe @slavapestov can.